File: learnLeetCode/Day29/Part3/dungeon_tokens.py
from typing import List

def max_tokens(grid: List[List[int]]) -> int:
    n = len(grid)
    if n == 0: return -1
    INF_NEG = -10**18
    dp = [[INF_NEG]*n for _ in range(n)]
    if grid[0][0] >= 0:
        dp[0][0] = grid[0][0]
    for i in range(n):
        for j in range(n):
            if i == 0 and j == 0:
                continue
            best = INF_NEG
            if i > 0 and dp[i-1][j] != INF_NEG and dp[i-1][j] + grid[i][j] >= 0:
                best = max(best, dp[i-1][j] + grid[i][j])
            if j > 0 and dp[i][j-1] != INF_NEG and dp[i][j-1] + grid[i][j] >= 0:
                best = max(best, dp[i][j-1] + grid[i][j])
            dp[i][j] = best
    return -1 if dp[n-1][n-1] == INF_NEG else dp[n-1][n-1]

if __name__ == "__main__":
    grid = [
        [1, -2, 3],
        [2, -1, 2],
        [1,  2, 1]
    ]
    print(max_tokens(grid))

File: learnLeetCode/Day29/Part3/dungeon_tokens.cpp
#include <bits/stdc++.h>
using namespace std;

long long max_tokens(const vector<vector<long long>>& grid){
    int n=grid.size(); if(!n) return -1; const long long INF_NEG = -(1LL<<60);
    vector<vector<long long>> dp(n, vector<long long>(n, INF_NEG));
    if(grid[0][0] >= 0) dp[0][0] = grid[0][0];
    for(int i=0;i<n;++i){
        for(int j=0;j<n;++j){
            if(i==0 && j==0) continue;
            long long best=INF_NEG;
            if(i>0 && dp[i-1][j]!=INF_NEG && dp[i-1][j]+grid[i][j] >= 0) best=max(best, dp[i-1][j]+grid[i][j]);
            if(j>0 && dp[i][j-1]!=INF_NEG && dp[i][j-1]+grid[i][j] >= 0) best=max(best, dp[i][j-1]+grid[i][j]);
            dp[i][j]=best;
        }
    }
    return dp[n-1][n-1]==INF_NEG? -1: dp[n-1][n-1];
}

int main(){
    vector<vector<long long>> g={{1,-2,3},{2,-1,2},{1,2,1}};
    cout<<max_tokens(g)<<"\n";
    return 0;
}
